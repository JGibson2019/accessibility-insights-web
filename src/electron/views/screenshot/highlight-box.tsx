// Copyright (c) Microsoft Corporation. All rights reserved.
// Licensed under the MIT License.
import { NamedFC } from 'common/react/named-fc';
import { BoundingRectangle } from 'electron/platform/android/scan-results';
import { highlightBox, highlightBoxLabel } from 'electron/views/screenshot/highlight-box.scss';
import * as React from 'react';
import { CSSProperties } from 'react';

export interface HighlightBoxProps {
    boundingRectangle: BoundingRectangle;
}

export const HighlightBox = NamedFC<HighlightBoxProps>('HighlightBox', props => {
    const { boundingRectangle } = props;

    const width = boundingRectangle.right - boundingRectangle.left;
    const height = boundingRectangle.bottom - boundingRectangle.top;

    const boxStyles: CSSProperties = {
        width: `${1.1 * width}px`,
        height: `${1.1 * height}px`,
        top: `${0.98 * boundingRectangle.top}px`,
        left: `${0.9 * boundingRectangle.left}px`,
    };

    return (
        <div className={highlightBox} style={boxStyles} ariaHidden="true">
            <div className={highlightBoxLabel}>!</div>
        </div>
    );
});
