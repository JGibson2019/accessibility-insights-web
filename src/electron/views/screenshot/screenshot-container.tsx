// Copyright (c) Microsoft Corporation. All rights reserved.
// Licensed under the MIT License.
import { NamedFC } from 'common/react/named-fc';
import { ScreenshotData } from 'common/types/store-data/unified-data-interface';
import { BoundingRectangle } from 'electron/platform/android/scan-results';
import { HighlightBox } from 'electron/views/screenshot/highlight-box';
import { Screenshot } from 'electron/views/screenshot/screenshot';
import { screenshotContainer } from 'electron/views/screenshot/screenshot-container.scss';
import { isEmpty } from 'lodash';
import * as React from 'react';

export interface ScreenshotContainerDeps {}
export interface ScreenshotContainerProps {
    deps?: ScreenshotContainerDeps;
    screenshotData: ScreenshotData;
    highlightBoxes?: BoundingRectangle[];
}

export const ScreenshotContainer = NamedFC<ScreenshotContainerProps>('ScreenshotContainer', props => {
    if (isEmpty(props.screenshotData)) {
        return <div>Screenshot for scan is unavailable</div>;
    }
    return (
        <div className={screenshotContainer}>
            <Screenshot encodedImage={props.screenshotData.base64PngData} />
            {renderHighlightBoxes(props.highlightBoxes)}
        </div>
    );
});

function renderHighlightBoxes(highlightBoxes: BoundingRectangle[]): JSX.Element[] {
    const renderedBoxes: JSX.Element[] = [];
    let keyId = 1;
    highlightBoxes.forEach(instance => {
        renderedBoxes.push(<HighlightBox key={keyId} boundingRectangle={instance} />);
        keyId++;
    });
    return renderedBoxes;
}
