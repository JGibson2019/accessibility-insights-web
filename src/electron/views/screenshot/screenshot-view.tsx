// Copyright (c) Microsoft Corporation. All rights reserved.
// Licensed under the MIT License.
import { NamedFC } from 'common/react/named-fc';
import { ScreenshotContainer } from 'electron/views/screenshot/screenshot-container';
import * as React from 'react';
import { ScreenshotViewModel } from './screenshot-view-model';

export type ScreenshotViewProps = { viewModel: ScreenshotViewModel };

export const ScreenshotView = NamedFC<ScreenshotViewProps>('ScreenshotView', (props: ScreenshotViewProps) => {
    return (
        <div>
            <h1 tabIndex={0}>Target Page Screenshot</h1>
            <ScreenshotContainer screenshotData={props.viewModel.screenshotData} />
        </div>
    );
});
