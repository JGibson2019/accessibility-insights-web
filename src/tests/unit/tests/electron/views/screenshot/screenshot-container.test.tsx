// Copyright (c) Microsoft Corporation. All rights reserved.
// Licensed under the MIT License.
import { ScreenshotData } from 'common/types/store-data/unified-data-interface';
import { BoundingRectangle } from 'electron/platform/android/scan-results';
import { ScreenshotContainer, ScreenshotContainerProps } from 'electron/views/screenshot/screenshot-container';
import { shallow } from 'enzyme';
import * as React from 'react';

describe('ScreenshotContainer', () => {
    const boundingRectangles: BoundingRectangle[] = [
        { top: 0, bottom: 100, left: 0, right: 100 },
        { top: 150, bottom: 200, left: 150, right: 300 },
    ];
    const basicScreenshotData: ScreenshotData = { base64PngData: 'test-base-64-png-data' };
    describe('renders', () => {
        it('when passed a value for screenshotData', () => {
            const props: ScreenshotContainerProps = {
                screenshotData: basicScreenshotData,
            };
            const wrapper = shallow(<ScreenshotContainer {...props} />);

            expect(wrapper.getElement()).toMatchSnapshot();
        });

        it('when passed an empty value for screenshotData', () => {
            const props = {
                screenshotData: {},
            } as ScreenshotContainerProps;

            const wrapper = shallow(<ScreenshotContainer {...props} />);

            expect(wrapper.getElement()).toMatchSnapshot();
        });

        it('when passed an null value for screenshotData', () => {
            const props = {
                screenshotData: null,
            } as ScreenshotContainerProps;

            const wrapper = shallow(<ScreenshotContainer {...props} />);

            expect(wrapper.getElement()).toMatchSnapshot();
        });

        it('when passed an undefined value for screenshotData', () => {
            const props = {
                screenshotData: undefined,
            } as ScreenshotContainerProps;

            const wrapper = shallow(<ScreenshotContainer {...props} />);

            expect(wrapper.getElement()).toMatchSnapshot();
        });
    });

    describe('highlight boxes', () => {
        it('do not render when boundingRectangles array is empty', () => {
            const props: ScreenshotContainerProps = {
                screenshotData: basicScreenshotData,
                highlightBoxes: [],
            };

            const wrapper = shallow(<ScreenshotContainer {...props} />);

            expect(wrapper.getElement()).toMatchSnapshot();
        });

        it('do not render when boundingRectangles array is not in props', () => {
            const props: ScreenshotContainerProps = {
                screenshotData: basicScreenshotData,
            };

            const wrapper = shallow(<ScreenshotContainer {...props} />);

            expect(wrapper.getElement()).toMatchSnapshot();
        });

        it('render when passed values for boundingRectangles array', () => {
            const props: ScreenshotContainerProps = {
                screenshotData: basicScreenshotData,
                highlightBoxes: boundingRectangles,
            };
            const wrapper = shallow(<ScreenshotContainer {...props} />);
            expect(wrapper.getElement()).toMatchSnapshot();
        });
    });
});
