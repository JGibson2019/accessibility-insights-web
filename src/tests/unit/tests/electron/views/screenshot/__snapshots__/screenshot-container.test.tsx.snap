// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ScreenshotContainer renders highlight boxes when passed values for boundingRectangles array 1`] = `
<div
  className="screenshotContainer"
>
  <Screenshot
    encodedImage="test-base-64-png-data"
  />
  <HighlightBox
    boundingRectangle={
      Object {
        "bottom": 100,
        "left": 0,
        "right": 100,
        "top": 0,
      }
    }
  />
  <HighlightBox
    boundingRectangle={
      Object {
        "bottom": 200,
        "left": 150,
        "right": 300,
        "top": 150,
      }
    }
  />
</div>
`;

exports[`ScreenshotContainer renders when passed a value for screenshotData 1`] = `
<div
  className="screenshotContainer"
>
  <Screenshot
    encodedImage="test-base-64-png-data"
  />
  <HighlightBox
    boundingRectangle={
      Object {
        "bottom": 100,
        "left": 0,
        "right": 100,
        "top": 0,
      }
    }
  />
  <HighlightBox
    boundingRectangle={
      Object {
        "bottom": 200,
        "left": 150,
        "right": 300,
        "top": 150,
      }
    }
  />
</div>
`;

exports[`ScreenshotContainer renders when passed an empty value for screenshotData 1`] = `
<div>
  Screenshot for scan is unavailable
</div>
`;

exports[`ScreenshotContainer renders when passed an null value for screenshotData 1`] = `
<div>
  Screenshot for scan is unavailable
</div>
`;

exports[`ScreenshotContainer renders when passed an undefined value for screenshotData 1`] = `
<div>
  Screenshot for scan is unavailable
</div>
`;
